### PR TITLE
Simplify the representation of `NullableWalker.LocalState` to a `BitVector`

### LINKED ISSUE #34766: Use BitVector for NullableWalker.LocalState
We should use `BitVector` in the implementation of `NullableWalker.LocalState`.  That will permit us to allocate and work with states without any memory allocation in the most common cases (fewer than 32 slots, but we could increase that to 64).

### PR DESCRIPTION
**Pull request by gafter**
Fixes #34766

Previously, every instance of `LocalState` required allocation.  Now we use `BitVector`, which stores state information inline in the bits of a primitive.

With the previous implementation of `BitVector` (32 bits inline), only about 1.3% of our nullable tests would require allocating any memory at all for nullable states.  However, I've updated `BitVector` to have 64 bits inline, and now only three of our tests require allocating memory for states.  Those three tests heavily exercise tuple conversions.

### CODE REVIEW

File: src/Compilers/Core/Portable/InternalUtilities/ArrayExtensions.cs [hunk 1]
```diff
@@ -47,6 +47,29 @@ internal static bool ValueEquals(this uint[] array, uint[] other)
...
+
+            for (int i = 0; i < array.Length; i++)
+            {
+                if (array[i] != other[i])
+                {
+                    return false;
+                }
+            }
+
+            return true;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/InternalUtilities/ArrayExtensions.cs]: Consider using Span for this.  IIRC, it has a vector optimized SequenceEquals check.

**Comment by cston** [comment]
>         internal static bool ValueEquals(this uint[] array, uint[] other)

It looks like this overload can be removed. The only caller was `BitVector`.

---
Refers to: src/Compilers/Core/Portable/InternalUtilities/ArrayExtensions.cs:27 in 354c141. [](commit_id = 354c141978816d88c4666572d744476b3571317d, deletion_comment = False)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -5905,103 +5880,63 @@ internal class LocalState : ILocalState
         internal struct LocalState : ILocalState
 #endif
         {
-            private ArrayBuilder<NullableFlowState> _state;
-            public bool Reachable { get; }
+            // The representation of a state is a bit vector.  We map false<->NotNull and true<->MayBeNull.
+            // Slot 0 is used to represent whether the state is reachable (true) or not.
+            private BitVector _state;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >private [](start = 12, length = 7)

`readonly`

File: src/Compilers/Core/Portable/InternalUtilities/ArrayExtensions.cs (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/InternalUtilities/ArrayExtensions.cs]: Essentially: `=> array.AsSpan().SequenceEquals(other.AsSpan());`

---
In reply to: [272446240](https://github.com/dotnet/roslyn/pull/34769#discussion_r272446240) [](ancestors = 272446240)

**Comment by gafter** [comment]
@cston @jaredpar I believe I have responded to your comments in the new iteration.  Do you have any other comments?

**Pull request merged**